### PR TITLE
DM-23080: dimension+registry refactoring

### PR DESCRIPTION
**Pull request by TallJimbo**

### CODE REVIEW

File: python/lsst/daf/butler/registry/dimensions/query.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+from ...core import DataCoordinate, Dimension, DimensionElement, DimensionRecord, Timespan
+from ...core.dimensions.schema import makeElementTableSpec
+from ...core.utils import NamedKeyDict
+from ..interfaces import Database, DimensionRecordStorage
+from ..queries import QueryBuilder
+
+
+class QueryDimensionRecordStorage(DimensionRecordStorage):
+    """A read-only record storage implementation that aggregates the values of
+    another element for which it is a required or implied dimension.
```
Comment (natelust) [inline python/lsst/daf/butler/registry/dimensions/query.py]: This comment could be made a little clearer

File: python/lsst/daf/butler/registry/dimensions/query.py [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+
+    Parameters
+    ----------
+    db : `Database`
+        Interface to the database engine namespace that will hold these
```
Comment (natelust) [inline python/lsst/daf/butler/registry/dimensions/query.py]: The use of the word namespace here is a little confusing

File: python/lsst/daf/butler/registry/dimensions/spatial.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+    Parameters
+    ----------
+    db : `Database`
+        Interface to the database engine namespace that will hold these
+        dimension records.
+    element : `DimensionElement`
+        The element whose records this storage will manage.
+    table : `sqlalchemy.schema.Table`
+        The logical table for the element.
+    commonSkyPixOverlapTable : `sqlalchemy.schema.Table`, optional
```
Comment (natelust) [inline python/lsst/daf/butler/registry/dimensions/spatial.py:81]: This does not seem optional in the init method

File: python/lsst/daf/butler/registry/dimensions/table.py [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+            [self._table.columns[name] for name in RecordClass.__slots__]
+        ).select_from(
+            self._table
+        ).where(sqlalchemy.sql.and_(*whereTerms))
+        row = self._db.query(query).fetchone()
+        if row is None:
+            return None
+        return RecordClass(*row)
+
+    def insert(self, *records: DimensionRecord):
```
Comment (natelust) [inline python/lsst/daf/butler/registry/dimensions/table.py:116]: So records is plural but its type is DimensionRecord. Is the type iterable or is the type hint wrong

File: python/lsst/daf/butler/registry/interfaces/_dimensions.py [hunk 5]
```diff
@@ -25,7 +25,7 @@
 from abc import ABC, abstractmethod
 from typing import Optional
 
-from sqlalchemy.sql import FromClause
+import sqlalchemy
```
Comment (natelust) [inline python/lsst/daf/butler/registry/interfaces/_dimensions.py:28]: seems like overkill to import everything. Why was this changed? and or why not just import sqlachemy.sql

File: python/lsst/daf/butler/registry/dimensions/spatial.py [hunk 6]
```diff
@@ -53,10 +56,18 @@ def __init__(self, db: Database, element: DimensionElement, *, table: sqlalchemy
...
-    def getCommonSkyPixOverlapTable(self, dataId: Optional[DataId] = None) -> sqlalchemy.schema.Table:
-        # Docstring inherited from
-        # DimensionRecordStorage.getCommonSkyPixOverlapTable.
-        return self._commonSkyPixOverlapTable
+    def join(
+        self,
+        builder: QueryBuilder, *,
+        regions: Optional[NamedKeyDict[DimensionElement, sqlalchemy.sql.ColumnElement]] = None,
+        timespans: Optional[NamedKeyDict[DimensionElement, Timespan[sqlalchemy.sql.ColumnElement]]] = None,
+    ):
```
Comment (natelust) [inline python/lsst/daf/butler/registry/dimensions/spatial.py:114]: I may not be following along with code flow, but what if this function gets called before one in which regions are added, should what is being added here get added some how?

File: python/lsst/daf/butler/registry/dimensions/table.py [hunk 7]
```diff
@@ -60,12 +62,25 @@ def clearCaches(self):
...
+    def join(
+        self,
+        builder: QueryBuilder, *,
+        regions: Optional[NamedKeyDict[DimensionElement, sqlalchemy.sql.ColumnElement]] = None,
+        timespans: Optional[NamedKeyDict[DimensionElement, Timespan[sqlalchemy.sql.ColumnElement]]] = None,
+    ):
+        assert regions is None, "This implementation does not handle spatial joins."
+        joinDimensions = list(self.element.graph.required)
+        joinDimensions.extend(self.element.implied)
+        joinOn = builder.startJoin(self._table, joinDimensions, self.element.RecordClass.__slots__)
```
Comment (natelust) [inline python/lsst/daf/butler/registry/dimensions/table.py:87]: Im slightly uncomfortable directly accessing __slots__ here, though I dont think it can hurt....

File: python/lsst/daf/butler/registry/queries/_builder.py [hunk 8]
```diff
@@ -94,50 +101,13 @@ def joinDimensionElement(self, element: DimensionElement):
...
-        if element is self.summary.universe.commonSkyPix:
-            return
-        assert element in self.summary.spatial,\
-            f"{element}'s spatial information is not relevant for this query."
-        table = self._dimensionStorage[element].getCommonSkyPixOverlapTable(self.summary.dataId)
-        assert table is not None, f"No relationship found from the common skypix to {element}."
-        dimensions = NamedValueSet(element.graph.required)
-        dimensions.add(self.summary.universe.commonSkyPix)
-        self.joinTable(table, dimensions)
+        storage = self._dimensionStorage[element]
```
Comment (natelust) [inline python/lsst/daf/butler/registry/queries/_builder.py:103]: do you want a normal lookup failure or something with a custom message if the element is not in the dimensionStorage

File: python/lsst/daf/butler/registry/interfaces/_dimensions.py [hunk 9]
```diff
@@ -48,6 +57,74 @@ class DimensionRecordStorage(ABC):
     subclasses to provide them is not a significant burden.
     """
 
+    @classmethod
+    @abstractmethod
+    def initialize(cls, db: Database, element: DimensionElement, *,
+                   context: Optional[StaticTablesContext] = None) -> DimensionRecordStorage:
+        """Construct an instance of this class using a standardized interface.
```
Comment (natelust) [inline python/lsst/daf/butler/registry/interfaces/_dimensions.py:64]: This is an odd docstring, because the words you used also exactly describe new/init

**Review by natelust — APPROVED** [review]
Some questions, otherwise good.

File: python/lsst/daf/butler/registry/dimensions/query.py (hunk 2 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/registry/dimensions/query.py]: I think I meant to say "engine and namespace" instead of "engine namespace". I am trying hard to use "namespace" instead of "schema" because the latter is even more overloaded.

File: python/lsst/daf/butler/registry/dimensions/table.py (hunk 4 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/registry/dimensions/table.py:116]: It's a variadic argument, and type annotations for variadic arguments (like those for the type yielded by a generator) are supposed to be for a single element.

File: python/lsst/daf/butler/registry/interfaces/_dimensions.py (hunk 5 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/registry/interfaces/_dimensions.py:28]: I've come to the conclusion that SQLAlchemy isn't well suited to `from <module> import <symbol>`, because so many of its symbols are tiny and don't provide enough context - things like `Sequence`, `Table`, `and_` etc.  More musings on this subject in a post from a few days ago [on slack](https://lsstc.slack.com/archives/C2JPL8U78/p1579878585048400).

And `import sqlalchemy.sql` would still invoke `sqlalchemy/__init__.py` - the only reason to use it would be to force the `sql` subpackage to also be imported, but SQLAlchemy is designed for that not to be necessary.

File: python/lsst/daf/butler/registry/dimensions/spatial.py (hunk 6 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/registry/dimensions/spatial.py:114]: The base class docs (looks like I need to add a comment pointing to them here) clarify that this method should only ever be called by `QueryBuilder.joinDimensionElement` (this is a visitor pattern thing), and it's responsible for guaranteeing that `regions` is provided (even if it is an empty dict, which is fine) only when appropriate.

File: python/lsst/daf/butler/registry/dimensions/table.py (hunk 7 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/registry/dimensions/table.py:87]: I've at least [documented](https://pipelines.lsst.io/v/daily/py-api/lsst.daf.butler.DimensionRecord.html#lsst.daf.butler.DimensionRecord) that `__slots__` is considered part of the public API for `DimensionRecord`, but I have always wondered whether it'd be better to add a differently-named class attribute with the exact same elements instead.  If you have a strong preference for that, this would not be a bad ticket to make that change.

File: python/lsst/daf/butler/registry/queries/_builder.py (hunk 8 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/registry/queries/_builder.py:103]: Normal failure is fine for now; I think this is guaranteed to succeed with the only `DimensionRecordStorageManager` implementation we have right now, and how to handle manager implementations that do not guarantee this (because they're in a multi-layer Registry in which only the base layer has tables for all dimensions) is sufficiently unsettled that I'm sure this line will have to change in other ways if we don't continue to guarantee its success.

File: python/lsst/daf/butler/registry/interfaces/_dimensions.py (hunk 9 above)
Comment (TallJimbo) [reply python/lsst/daf/butler/registry/interfaces/_dimensions.py:64]: Not quite - `__new__` (which is what this is more like) is not usually expected to have the same signature (what I mean by "standardized") across all subclasses of an ABC, and that's also what makes `__new__` inappropriate (and maybe disallowed?) as an `abstractmethod`.
